src/components/metadata.cairo

```diff
@@ -0,0 +1,109 @@
+use starknet::ClassHash;
+
+#[starknet::interface]
+trait IMetadataHandler<TContractState> {
+    fn uri(self: @TContractState, token_id: u256) -> ByteArray;
+    fn set_uri(ref self: TContractState, class_hash: ClassHash);
+}
+
+#[starknet::interface]
+trait IMetadataDescriptor<TContractState> {
+    fn construct_uri(self: @TContractState, token_id: u256) -> ByteArray;
+}
+
+#[starknet::component]
+mod metadata_component {
+    use starknet::ClassHash;
+    use super::{IMetadataDescriptorLibraryDispatcher, IMetadataDescriptorDispatcherTrait};
+
+    #[storage]
+    struct Storage {
+        uri_implementation: ClassHash
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct MetadataUpgraded {
+        old_class_hash: ClassHash,
+        class_hash: ClassHash,
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    pub enum Event {
+        MetadataUpgraded: MetadataUpgraded,
+    }
+
+    #[embeddable_as(CarbonV3Metadata)]
+    impl MetadataImpl<
```

Comment:
The convention from OpenZepplin would be :
```
    #[embeddable_as(CarbonV3MetadataImpl)]
    impl CarbonV3Metadata<
```

Like you can see in other components